oscarlab/graphene
# Loader rework part 5: central manifest and constant MRENCLAVE

## Description of the changes 

// Sorry for such a big change in a single commit (and PR), but I really don't have a good idea how to split it into smaller steps.

This is the next part of the great loader rework (previously in this series: #1895, #1901, #1915, #2008), with a lot of breaking changes:

- Complete removal of the "trusted children" thing - now children processes can be spawned arbitrarily and from arbitrary mountpoint types, without any additional configuration needed.
- There's a new, required option in the manifest: `libos.entrypoint` - it specifies the URI to the entry binary in the first process. There's no need anymore to name the manifest and the first binary identically.
- On SGX, the main binary is not measured in MRENCLAVE anymore - only PAL, LibOS and the manifest are measured. This is enough to bind MRENCLAVE to the specific entrypoint user executable if wanted - it just has to be mounted as a trusted file.
- All Graphene SGX enclaves have now exactly the same MRENCLAVE. This is a hash of a "Graphene stub", which can "fork" into one of two states in runtime: initial process or child. The initial process creates a new "Graphene namespace" with a clean state, it can also be attested remotely (contrary to child processes). The initial process can spawn children processes by spawning a Graphene stub and directing it to start in the child mode. It then attests it locally, and if successful, establishes an encrypted pipe, "connects" to its own namespace and treats as trusted (including sending protected files key).
- Now, there's only one, central manifest describing the initial state of a Graphene instance which can be spawned from it (previously, each process required a separate manifest which could have different configuration - which wasn't actually supported and didn't make sense design-wise). One downside of central manifests is that all processes require the same enclave configuration (e.g. size), but that was already the case so far because of broken checkpointing code. Also, this is only a temporary problem, which will cease to exist after the introduction of EDMM.
- `sgx.static_address` was renamed to `sgx.nonpie_binary` and now has to be inserted manually by users (`sgx_sign` tools doesn't know about the binaries run inside, which can be even provided or generated in runtime by the user's workload).
- Caveat: the memory gap for non-PIE executables was removed because it requires adding a new option to the manifest to be cleanly implemented. This is left for some future loader rework PR.


Next plans:

- pal_loader's argument should (IMO) now point to the manifest, not the "project name". It was added as a temporary measure to keep the old invocation syntax for now (this PR introduces a lot of changes already).
- PAL loader code should have no notion of the entrypoint binary - this should be handled solely by the LibOS.
- We need to add a manifest option to reserve user-chosen memory address ranges for non-PIE executables. Upon trying of loading a non-PIE executable without such a reservation, we'd print an error with a snippet which has to be added to the manifest (when printing it we'll already know the exact range used).
- We need to double check (and block in software) that accidental remote attestation of children is not possible (only the initial process should be remotely attested).
- Add separate tests for all weird loading patterns (e.g. starting from a protected file).

Questions/doubts/bugs:

- I didn't do a proper "garbage collecting" stage yet, so you may find some now-unused code while reviewing, please point such cases out. I'll do some cleanups after the next PR (see above).
- GCC example still doesn't work, one of the children is crashing. I left debugging this for later, as everything else works.
- After my changes a weird bug in Make started to appear indeterministically: sometimes, after I change a few files, building LibOS regression paralelly overwrites Scripts/Makefile.manifest (using manifest.template from LibOS regression dir). I have no idea how this is possible and I couldn't find which rule/include causes its regeneration, nor how to reproduce it reliably.
- I gave up fixing GSC, too PITA to test changes :/ I think we should move it to a separate repository in our GitHub org, so that both (Graphene and GSC can be updated independently). CC: @vahldiek
- Kudos to @dimakuv for debugging a bug in RTLD code which I triggered :)

Resolves #972. May also help with #1989, but this needs to be verified.


## How to test this PR? 

Current tests should cover most of the cases. I plan to add more tests with weird cases, but rather after finishing the changes from the "Next plans" above.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene/2050)
